File: src/com/jkbff/ao/tyrlib/chat/ChatPacketSender.java
package com.jkbff.ao.tyrlib.chat;

import java.io.IOException;
import java.io.OutputStream;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

import org.apache.log4j.Logger;

import com.jkbff.ao.tyrlib.packets.BaseClientPacket;

public class ChatPacketSender extends Thread {

	private OutputStream outputStream;
	private BlockingQueue<BaseClientPacket> packetQueue = new ArrayBlockingQueue<BaseClientPacket>(1000);
	private AOSingleConnection aoBot;
	private Logger log = Logger.getLogger(this.getClass());

	@Override
	public void run() {
        while (!aoBot.shouldStop) {
            try {
                BaseClientPacket packet = packetQueue.poll(1, TimeUnit.SECONDS);
                if (packet != null) {
                	byte[] bytes = packet.getBytes();
                	outputStream.write(bytes);
                }
            } catch (InterruptedException e) {
            	log.error(e);
            } catch (IOException e) {
            	log.error(e);
            }
        }

        try {
            outputStream.close();
        } catch (Exception e) {
        	log.error(e);
        }
	}

	public void sendPacket(BaseClientPacket packet) {
		packetQueue.add(packet);
	}

	public void setOutputStream(OutputStream outputStream) { this.outputStream = outputStream; }
	public void setAOBot(AOSingleConnection aoBot) { this.aoBot = aoBot; }
}

File: src/main/scala/com/jkbff/ao/tyrlib/chat/ChatPacketSender.scala
package com.jkbff.ao.tyrlib.chat
import java.io.OutputStream
import org.apache.log4j.Logger
import java.util.concurrent.LinkedBlockingQueue
import com.jkbff.ao.tyrlib.packets.BaseClientPacket
import com.jkbff.ao.tyrlib.packets.client.PrivateMessageSend
import com.jkbff.ao.tyrlib.packets.client.ChannelMessage

class ChatPacketSender(outputStream: OutputStream, aoBot: AOSocket) extends Thread {
	val log = Logger.getLogger(this.getClass());

	override def run() {
		while (!aoBot.shouldStop) {
			try {
				val packet = aoBot.getNextPacketToSend()
				if (packet != null) {
					val bytes = packet.getBytes()
					outputStream.write(bytes)
				}
			} catch {
				case e: Exception => log.error("", e)
			}
		}

		try {
			outputStream.close()
		} catch {
			case e: Exception => log.error("", e)
		}
	}
}